infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java

```diff
@@ -131,69 +141,130 @@ public LogUnitServer(ServerContext serverContext) {
     public void write(CorfuPayloadMsg<WriteRequest> msg, ChannelHandlerContext ctx, IServerRouter r) {
         log.debug("log write: global: {}, streams: {}", msg.getPayload().getToken(),
                 msg.getPayload().getData().getBackpointerMap());
-
-        try {
-            LogData logData = (LogData) msg.getPayload().getData();
-            logData.setEpoch(msg.getEpoch());
-            dataCache.put(msg.getPayload().getGlobalAddress(), logData);
-            r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
-
-        } catch (OverwriteException ex) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE.payloadMsg(ex.getOverWriteCause().getId()));
-        } catch (DataOutrankedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } catch (ValueAdoptedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(e
-                    .getReadResponse()));
-        }
+        LogData logData = (LogData) msg.getPayload().getData();
+
+        batchWriter.addTask(WRITE, msg)
+                .thenAccept(Void -> {
+                    dataCache.put(msg.getPayload().getGlobalAddress(), logData);
+                    r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
+                }).exceptionally(ex -> {
+                    handleException(ex, ctx, msg, r);
+                    return null;
+        });
     }
 
     /**
      * Services incoming range write calls.
      */
     @ServerHandler(type = CorfuMsgType.RANGE_WRITE)
     public void rangeWrite(CorfuPayloadMsg<RangeWriteMsg> msg,
-                            ChannelHandlerContext ctx, IServerRouter r) {
-        try {
-            List<LogData> entries = msg.getPayload().getEntries();
-            batchWriter.bulkWrite(entries, msg.getEpoch());
-            r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg());
-        } catch (OverwriteException ex) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_OVERWRITE.payloadMsg(ex.getOverWriteCause().getId()));
-        } catch (DataOutrankedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_DATA_OUTRANKED.msg());
-        } catch (ValueAdoptedException e) {
-            r.sendResponse(ctx, msg, CorfuMsgType.ERROR_VALUE_ADOPTED.payloadMsg(e
-                    .getReadResponse()));
-        }
+                           ChannelHandlerContext ctx, IServerRouter r) {
+        List<LogData> range = msg.getPayload().getEntries();
+        log.debug("rangeWrite: Writing {} entries [{}-{}]", range.size(),
+                range.get(0).getGlobalAddress(), range.get(range.size() - 1).getGlobalAddress());
+
+        batchWriter.addTask(RANGE_WRITE, msg)
+                .thenAccept(Void -> r.sendResponse(ctx, msg, CorfuMsgType.WRITE_OK.msg()))
+                .exceptionally(ex -> {
+                    handleException(ex, ctx, msg, r);
+                    return null;
```

Comment:
I don't like that we have to return null. Can you please replace exceptionally() by handle(...) or whenCompleted(...)